// eslint-plugin-playwright/lib/index.js
"use strict";
const missing_playwright_await_1 = require("./rules/missing-playwright-await");
const no_page_pause_1 = require("./rules/no-page-pause");
const no_element_handle_1 = require("./rules/no-element-handle");
const no_eval_1 = require("./rules/no-eval");
const no_focused_test_1 = require("./rules/no-focused-test");
const no_skipped_test_1 = require("./rules/no-skipped-test");
const no_wait_for_timeout_1 = require("./rules/no-wait-for-timeout");
const no_force_option_1 = require("./rules/no-force-option");
const max_nested_describe_1 = require("./rules/max-nested-describe");
const no_conditional_in_test_1 = require("./rules/no-conditional-in-test");
const no_restricted_matchers_1 = require("./rules/no-restricted-matchers");
const no_useless_not_1 = require("./rules/no-useless-not");
const prefer_lowercase_title_1 = require("./rules/prefer-lowercase-title");
const prefer_to_have_length_1 = require("./rules/prefer-to-have-length");
const require_top_level_describe_1 = require("./rules/require-top-level-describe");
const valid_expect_1 = require("./rules/valid-expect");
module.exports = {
    configs: {
        'playwright-test': {
            plugins: ['playwright'],
            env: {
                'shared-node-browser': true,
            },
            rules: {
                'no-empty-pattern': 'off',
                'playwright/missing-playwright-await': 'error',
                'playwright/no-page-pause': 'warn',
                'playwright/no-element-handle': 'warn',
                'playwright/no-eval': 'warn',
                'playwright/no-focused-test': 'error',
                'playwright/no-skipped-test': 'warn',
                'playwright/no-wait-for-timeout': 'warn',
                'playwright/no-force-option': 'warn',
                'playwright/max-nested-describe': 'warn',
                'playwright/no-conditional-in-test': 'warn',
                'playwright/no-useless-not': 'warn',
                'playwright/valid-expect': 'error',
            },
        },
        'jest-playwright': {
            plugins: ['jest', 'playwright'],
            env: {
                'shared-node-browser': true,
                jest: true,
            },
            rules: {
                'playwright/missing-playwright-await': 'error',
                'playwright/no-page-pause': 'warn',
                'jest/no-standalone-expect': [
                    'error',
                    {
                        additionalTestBlockFunctions: [
                            'test.jestPlaywrightDebug',
                            'it.jestPlaywrightDebug',
                            'test.jestPlaywrightSkip',
                            'it.jestPlaywrightSkip',
                            'test.jestPlaywrightConfig',
                            'it.jestPlaywrightConfig',
                        ],
                    },
                ],
            },
            globals: {
                browserName: true,
                deviceName: true,
                browser: true,
                context: true,
                page: true,
                jestPlaywright: true,
            },
        },
    },
    rules: {
        'missing-playwright-await': missing_playwright_await_1.default,
        'no-page-pause': no_page_pause_1.default,
        'no-element-handle': no_element_handle_1.default,
        'no-eval': no_eval_1.default,
        'no-focused-test': no_focused_test_1.default,
        'no-skipped-test': no_skipped_test_1.default,
        'no-wait-for-timeout': no_wait_for_timeout_1.default,
        'no-force-option': no_force_option_1.default,
        'max-nested-describe': max_nested_describe_1.default,
        'no-conditional-in-test': no_conditional_in_test_1.default,
        'no-useless-not': no_useless_not_1.default,
        'no-restricted-matchers': no_restricted_matchers_1.default,
        'prefer-lowercase-title': prefer_lowercase_title_1.default,
        'prefer-to-have-length': prefer_to_have_length_1.default,
        'require-top-level-describe': require_top_level_describe_1.default,
        'valid-expect': valid_expect_1.default,
    },
};
